### DISCUSSION TITLE
Whats the proper way to update a uniform for each mesh?

### DISCUSSION THREAD

**Discussion by cdecompilador** [gfx-rs/wgpu#3244] [Q&A] [answered]
At the moment if I want to draw two different meshes that have two different vertex shader and index shaders I have to create an independent `CommandEncoder` and `RenderPass` for each of them, since the `write_buffer` is only effective one you do `queue.submit()`, is this the proper way to render two different meshes with different bind groups or you can do it all on a single render pass?

Thanks in advance.

**Comment by kpreid** [comment, marked as answer]
You can create two or more uniform buffers — one for each set of uniform values you want to use during that frame. You will also need a bind group for each buffer.

There are other options too, like storing arrays of values in a single buffer, which can then be indexed using attributes in the mesh data, or the instance number. Which one is best depends on exactly what you're drawing.

The way to think about the problem in general is: there should be places somewhere in your buffers (and textures) for *all the data you need for one frame, all at the same time*, so that the order of operations can be thought of as “write, write, draw, draw” rather than “write, draw, write, draw”. There are various ways you could organize the data, but all the data should be conceptually present simultaneously.
